### PR TITLE
Add DD tags to log instrumentations

### PR DESCRIPTION
**Pull request by lpriima**
1. revert of this revert : https://github.com/DataDog/dd-trace-java/pull/1662revert of this revert : https://github.com/DataDog/dd-trace-java/pull/1662
1. fix for https://github.com/DataDog/dd-trace-java/issues/1654 and test for it: 

The issue https://github.com/DataDog/dd-trace-java/issues/1654 was introduced after instrumentation of log4j2. Original instrumentation assumed that interface ThreadContextMap always contains threadlocal field with instance of Map in generic parameter.
```
ThreadLocal<Map<?, ?>> localMap
```

In fact, log4j2 has 2 other implementations ( [CopyOnWriteSortedArrayThreadContextMap](https://github.com/apache/logging-log4j2/blob/master/log4j-api/src/main/java/org/apache/logging/log4j/spi/CopyOnWriteSortedArrayThreadContextMap.java) , [GarbageFreeSortedArrayThreadContextMap](https://github.com/apache/logging-log4j2/blob/master/log4j-api/src/main/java/org/apache/logging/log4j/spi/GarbageFreeSortedArrayThreadContextMap.java)  ) which contains  StringMap in generic parameter:
```
ThreadLocal<org.apache.logging.log4j.util.StringMap> localMap
```
And this org.apache.logging.log4j.util.StringMap is not instance of java.util.Map. And It was causing ClassCastException in https://github.com/DataDog/dd-trace-java/issues/1654 .

As a solution in this changeset class [ThreadLocalWithDDTagsInitValue<T> extends ThreadLocal<T>](https://github.com/DataDog/dd-trace-java/pull/1688/files#diff-c0f391488f8a9b254794c9f452e9756fR13) has also generic parameter, which will contain "Map-like" type of instrumented implementation. "Map-like" property means method  [ThreadLocalWithDDTagsInitValue#putToMap](https://github.com/DataDog/dd-trace-java/pull/1688/files#diff-c0f391488f8a9b254794c9f452e9756fR41) looks for the way how to "put" datadog key/values in the instance of this type by following  [duck typing](https://en.wikipedia.org/wiki/Duck_typing) scenario:
1. if this type is instance of Map, then use Map#put(String,String) method. If Not
1. if this type has `put(String, ?)` method then use it. If not
1. if this type has `putValue(String, ?)` method then use it. If not
1. skip insertion of DD tags with WARN log message

Regression test [Log4jThreadContextWithEnableThreadLocalsTest](  https://github.com/DataDog/dd-trace-java/pull/1688/files#diff-024eafb70a73f1931f9ec01bcd0e9992R7 ) for this issue, modifies static final field to ensure log4j2 uses org.apache.logging.log4j.util.StringMap.

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java [hunk 1]
```diff
@@ -204,6 +204,9 @@ private static void deleteTempDir(final File file) {
     final boolean deleted = file.delete();
     if (!deleted) {
       file.deleteOnExit();
+      log.debug("file '{}' added to shutdown delete hook", file);
+    } else {
+      log.debug("file '{}' deleted", file);
```
Comment (lpriima) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java:209]: This logging is not related to this change

**Comment by lpriima** [comment]
> Can you identify which tests were added to validate which previous problems were solved?

Log4jThreadContextWithEnableThreadLocalsTest that's bullet 2 of description

**Comment by tylerbenson** [comment]
It isn't clear to me how that validates the fix.  Please clarify.

**Comment by lpriima** [comment]
> It isn't clear to me how that validates the fix. Please clarify.
> 
> Perhaps better describe the problem this is solving and how this fixes it?

I've updates description here. 
Should I also put it in the issue https://github.com/DataDog/dd-trace-java/issues/1654 comments ?

File: dd-java-agent/agent-tooling/src/test/groovy/datadog/trace/agent/test/ClassLoaderMatcherTest.groovy [hunk 2]
```diff
@@ -38,6 +40,12 @@ class ClassLoaderMatcherTest extends DDSpecification {
     DatadogClassLoader.name == "datadog.trace.bootstrap.DatadogClassLoader"
   }
 
+  def "helper class names are hardcoded in Log Instrumentations"() {
+    expect:
+    LogContextScopeListener.name == "datadog.trace.agent.tooling.log.LogContextScopeListener"
+    ThreadLocalWithDDTagsInitValue.name == "datadog.trace.agent.tooling.log.ThreadLocalWithDDTagsInitValue"
+  }
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/agent-tooling/src/test/groovy/datadog/trace/agent/test/ClassLoaderMatcherTest.groovy:47]: I would assume the purpose is to guard against automated refactoring changing the classes but not the class names hard coded in various places. i.e this test would fail if either of these classes were renamed.

[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/test/groovy/datadog/trace/agent/test/ClassLoaderMatcherTest.groovy:47]: I believe most refactoring tools would actually detect this usage and also update the string since it's a fully qualified class name.

[the remark this answers is not in the document]
Comment (lpriima) [reply dd-java-agent/agent-tooling/src/test/groovy/datadog/trace/agent/test/ClassLoaderMatcherTest.groovy:47]: Yes, this is for possible class renaming and it's temporary solution.
Better solution should be running tests with actual classpath restricted by this method:
https://github.com/DataDog/dd-trace-java/pull/1688/files#diff-ddc855d905a56896d1c42dfd404e8541R50-R53
